---
title: "Analysis of simulated genomic data using a RDA-based genome scan approach"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##
## Libraries 

```{r, message=FALSE}
library("vegan")
library("robust")
library("qvalue")
library("ggplot2")
```

## Genetic dataset and environmental variables

```{r, message=FALSE}
geno<-read.csv("sim1.csv")[,-c(1:14)]
env<-read.csv("sim1.csv")[,c(1:14)]
```

## Removing loci with freq < MAF (Minor Allele Frequency)

```{r, message=FALSE}
MAF <- 0.05
frequencies <- colSums(geno)/(2*nrow(geno))
maf <- which(frequencies > MAF & frequencies < (1-MAF))
geno <- geno[,maf]
```

## Running the RDA

```{r, message=FALSE}
RDA <- rda(geno ~ env$envir1 + env$envir2 + env$envir3 + env$envir4 + env$envir5 + env$envir6 + env$envir7 + env$envir8 + env$envir9 + env$envir10,  env)
```

## Post treatment of RDA results

```{r, message=FALSE}
# The function returning p-values and q-values for all the loci

rdadapt<-function(rda,K)
{
  loadings<-rda$CCA$v[,1:as.numeric(K)]
  resscale <- apply(loadings, 2, scale)
  resmaha <- covRob(resscale, distance = TRUE, na.action= na.omit, estim="pairwiseGK")$dist
  lambda <- median(resmaha)/qchisq(0.5,df=K)
  reschi2test <- pchisq(resmaha/lambda,K,lower.tail=FALSE)
  qval <- qvalue(reschi2test)
  q.values_rdadapt<-qval$qvalues
  return(data.frame(p.values=reschi2test, q.values=q.values_rdadapt))
}

```
```{r, message=FALSE}
# Chosing the best number of axes for the analysis
```
```{r, echo=TRUE, fig.align="center", fig.width=4.5, fig.height=4.5}

ggplot() +
  geom_line(aes(x=c(1:length(RDA$CCA$eig)), y=as.vector(RDA$CCA$eig)), linetype="dotted", size = 1.5, color="darkgrey") +
  geom_point(aes(x=c(1:length(RDA$CCA$eig)), y=as.vector(RDA$CCA$eig)), size = 3, color="darkgrey") +
  scale_x_discrete(name = "Ordination axes", limits=c(1:9)) +
  ylab("Inertia") +
  theme_bw()
```
```{r, message=FALSE}
# Using rdadapt function with K=5 

res_rdadapt<-rdadapt(RDA, 5)
```

## Visualisation of the results

```{r, message=FALSE}
# Manhattan plot (outliers : q.value < 0.1 are colored in orange)
```
```{r, echo=TRUE, fig.align="center", fig.width=4.5, fig.height=4.5}
ggplot() +
  geom_point(aes(x=c(1:length(res_rdadapt[,1])), y=-log10(res_rdadapt[,1])), col = "gray83") +
  geom_point(aes(x=c(1:length(res_rdadapt[,1]))[which(res_rdadapt[,2] < 0.1)], y=-log10(res_rdadapt[which(res_rdadapt[,2] < 0.1),1])), col = "orange") +
  xlab("SNPs") + ylab("-log10(p.values)") +
  theme_bw()
```
##
```{r, message=FALSE}
# Outliers loci (q.values < 0.1)

which(res_rdadapt[,2] < 0.1)
```
##
```{r, message=FALSE}
# Projection of loci in the RDA space
```
```{r, echo=TRUE, fig.align="center", fig.width=4.5, fig.height=4.5}
ggplot() +
  geom_point(aes(x=RDA$CCA$v[,1], y=RDA$CCA$v[,2]), col = "gray86") +
  geom_point(aes(x=RDA$CCA$v[which(res_rdadapt[,2] < 0.1),1], y=RDA$CCA$v[which(res_rdadapt[,2] < 0.1),2]), col = "orange") +
  geom_segment(aes(xend=RDA$CCA$biplot[,1]/10, yend=RDA$CCA$biplot[,2]/10, x=0, y=0), colour="black", size=0.5, linetype=1, arrow=arrow(length = unit(0.02, "npc"))) +
  geom_text(aes(x=1.2*RDA$CCA$biplot[,1]/10, y=1.2*RDA$CCA$biplot[,2]/10, label = colnames(env[,2:11]))) +
  xlab("RDA 1") + ylab("RDA 2") +
  theme_bw() +
  theme(legend.position="none")

```
